src/code/z_demo.c

```diff
@@ -971,144 +1003,145 @@ s32 Cutscene_CountNormalMasks(void) {
     return count;
 }
 
-// Command 0xA: Textbox
-void Cutscene_Command_Textbox(PlayState* play, CutsceneContext* csCtx, CsCmdTextbox* cmd) {
-    static s32 D_801BB160 = CS_TEXTBOX_TYPE_DEFAULT;
+void CutsceneCmd_Text(PlayState* play, CutsceneContext* csCtx, CsCmdText* cmd) {
+    static s32 sCutsceneTextboxType = CS_TEXT_TYPE_DEFAULT;
     u8 talkState;
     s32 pad;
-    u16 originalCsFrames;
-    s32 pad2;
+    u16 endFrame;
 
-    if ((cmd->startFrame >= csCtx->frames) || ((cmd->endFrame < csCtx->frames))) {
+    if ((cmd->startFrame >= csCtx->curFrame) || ((cmd->endFrame < csCtx->curFrame))) {
         return;
     }
 
-    if (cmd->type != CS_TEXTBOX_TYPE_LEARN_SONG) {
-        if (D_801BB124 != cmd->base) {
-            if (D_801BB160 == CS_TEXTBOX_TYPE_3) {
-                csCtx->frames--;
+    if (cmd->type != CS_TEXT_OCARINA_ACTION) {
+        if (sCurTextId != cmd->textId) {
```

Comment:
```suggestion
    if (cmd->type != CS_TEXT_OCARINA_ACTION) {
        if (sCurTextId == cmd->textId) {
            // reintroduce else_label
            goto message_is_active; //or a better label name;
        }
```

It's a bit unusual, but I feel that adding the goto back here would actually make the code more readable since it allows us to express how the final code block (the old `else_label`) is reached logically. You can also do the same inversion in the `cmd->type == CS_TEXT_OCARINA_ACTION` block. Unfortunately it would mean having two gotos, but I think it's something to consider.